lib/generate-types.js

```diff
@@ -0,0 +1,132 @@
+      structsByNamespace.set(ns, [])
+    }
+    structsByNamespace.get(ns).push(struct)
+  }
+
+  let typings = ''
+  typings += `// Auto-generated by hyperschema codegen\n\n`
+
+  // Generate namespaces
+  for (const [namespace, structs] of structsByNamespace.entries()) {
+    typings += `export namespace ${sanitize(namespace)} {\n`
+
+    for (const struct of structs) {
+      if (struct.isEnum) {
+        typings += `  export enum ${sanitize(struct.name)} {\n`
+        for (const e of struct.enum) {
+          const key = e.key.replace(/[^a-zA-Z0-9_]/g, '_')
+          typings += `    ${sanitize(key)} = ${JSON.stringify(e.key)},\n`
+        }
+        typings += '  }\n\n'
+        continue
+      }
+
+      if (struct.isArray) {
+        const typeName = sanitize(struct.name)
+        const elementType = tsType(struct.type, struct.namespace)
+        typings += `  export type ${typeName} = ${elementType}[];\n\n`
+        continue
+      }
+
+      if (struct.isStruct || struct.isVersioned) {
+        const typeName = sanitize(struct.name)
+        typings += `  export interface ${typeName} {\n`
+
+        if (struct.isVersioned) typings += `    version?: number;\n`
+
+        for (const field of struct.fields || []) {
+          const name = field.name
+          const optional = field.required ? '' : '?'
+          const safeName = quoteIfNeeded(name)
+
+          typings += `    ${safeName}${optional}: ${tsType(field.type, struct.namespace)};\n`
+        }
+
+        typings += '  }\n\n'
+      }
+    }
+
+    typings += '}\n\n' // close namespace
+  }
+
+  // Global encode/decode API
+  typings += `
+export function encode<T>(name: string, value: T, v?: number): Uint8Array;
+export function decode<T>(name: string, buffer: Uint8Array, v?: number): T;
+export function getEncoding(name: string): any;
+export function getEnum(name: string): any;
+export function getStruct(name: string, v?: number): any;
+export function setVersion(v: number): void;
+export const version: number;
+`
+
+  function quoteIfNeeded(name) {
+    if (/^[A-Za-z_$][A-Za-z0-9_$]*$/.test(name)) return name
+    return `'${name.replace(/'/g, "\\'")}'`
+  }
+
+  function tsType(type, currentNamespace) {
+    if (!type) return 'any'
+
+    if (type.isAlias) return tsType(type.type, currentNamespace)
+
+    if (type.isPrimitive) {
+      const t = type.name
+
+      if (NumericTypes.has(t)) return 'number'
+      if (BigIntTypes.has(t)) return 'bigint'
+      if (BooleanTypes.has(t)) return 'boolean'
+      if (StringTypes.has(t)) return 'string'
+      if (NetworkTypes.has(t)) return 'string'
+      if (t === 'json') return 'any'
+
+      if (ObjectTypes.has(t)) {
+        switch (t) {
+          case 'buffer':
+          case 'fixed32':
+          case 'fixed64':
+            return 'Uint8Array'
+          case 'date':
+            return 'Date'
+        }
+      }
+
+      return 'any'
+    }
+
+    const targetNs = type.namespace || null
```

Comment:
When would we not have a `type.namespace` here? is that just the case for primitives?